chore: Write README for tokens package (#1655)
Amsterdam/design-system: proprietary/tokens/README.md

Every token has a value that is valid in CSS.

Some examples:

```css
:root {
  --ams-color-blue: #004699;
  --ams-text-font-weight-bold: 800;
  --ams-border-width-medium: 0.125rem;
  --ams-button-padding-inline: 1.5rem;
  --ams-space-xl: clamp(2.25rem, 2.0625rem + 0.9375vw, 3rem);
  --ams-text-input-invalid-border-color: var(--ams-color-red);
}
```

All tokens together form a theme that encodes the entire branding of the City of Amsterdam.

We use these tokens in our [CSS components](https://www.npmjs.com/package/@amsterdam/design-system-css) and [React components](https://www.npmjs.com/package/@amsterdam/design-system-react).
Any web-based application for Amsterdam can implement the design system through this package.

## Install the package

Use this package alone if you want or need to use the design tokens directly in your project.
Note that our [CSS components](https://www.npmjs.com/package/@amsterdam/design-system-css) and [React components](https://www.npmjs.com/package/@amsterdam/design-system-react) provide more functionality and apply the tokens automatically.

```sh

> pnpm add @amsterdam/design-system-tokens

Reviewer: Npm is more conventional, no?

```suggestion
npm i @amsterdam/design-system-tokens
```

```

## Import the tokens

Find the tokens in the `dist` directory and import them into your project.

> Tokens are typically applied as custom properties in CSS.

Reviewer: ```suggestion
Tokens are typically used as custom properties in CSS.
```

```ts
import "@amsterdam/design-system-tokens/dist/index.css";
```

### Root or class selector

The stylesheets wrap the tokens into a `:root` selector.
If that doesn’t work for your project, use `index.theme.css` instead, which uses an `.ams-theme` selector.
Add that class to a root element of your application.

### Compact mode

Applications need less whitespace and smaller text sizes than websites.
Add the compact stylesheet to override the values of these tokens.

```ts
import "@amsterdam/design-system-tokens/dist/index.css";
import "@amsterdam/design-system-tokens/dist/compact.css";
```

> Dark mode will work the same way when it’s ready.

Reviewer: It's a little odd to mention something we don't have, I think. And it will probably work the same way, but we don't know that for sure yet.

### Other formats

The tokens can be imported as Sass variables as well.

```sass
@import "@amsterdam/design-system-tokens/dist/index.scss"
```

To use the tokens in JavaScript, import the JSON file.

```ts
import tokens from "@amsterdam/design-system-tokens/dist/index.json";
```

> ## Use the tokens

Reviewer: This section is CSS specific again, no? Maybe we should specify that.

```suggestion
## Use the CSS tokens
```

Author: Thanks, this made me restructure the entire piece 😎

All concepts (three layers, compact mode etc.) are now explained under ‘CSS’, followed by small sections on secondary usage in Sass and JavaScript.

The latter made me wonder whether we should still export the `.mjs` format – JSON seems easier to use.

Reviewer: I think Robbert was working on a generic list of export types of tokens for the entire NLDS community, so I wouldn't mess too much with it now.

All tokens' names start with a prefix of `ams-`.
The name of component tokens ends with the CSS property that uses them, e.g. `-font-size`.

### Component tokens

Every component uses tokens for every property that expresses branding.
Apply these values in your stylesheets to the corresponding components.

```html
<button class="my-button" type="button">Button label</button>
```

```css
.my-button {
  font-family: var(--ams-button-font-family);
  background-color: var(--ams-button-primary-background-color);
}
```

### Common tokens

Various groups of components share visual design characteristics.
E.g. all kinds of links have the same colour, and the borders of various inputs are equally thick.
We use common tokens to express these relations.
This also makes it easier to change this shared appearance in the future.

Prefer using common tokens over component tokens.

> If

Reviewer: This looks accidental

```html
<input class="my-input" type="text" />
```

```css
.my-input {
  font-size: var(--ams-common-input-font-size);

>   border-width: var(--ams-common-input-border-width);

Reviewer: These are none-existent tokens, no? I would avoid that, to prevent copy paste frustration.

Author: I know but they do illustrate the example better than other token names. And I intend to adopt this naming strategy for brand and common tokens. I’ve added a notice that these don’t exist yet.

}
```

### Brand tokens

The brand tokens express the branding of the City of Amsterdam.
They are our selection from all possible colours, text characteristics, spacing lengths, border widths, etc.

> They are the basis of everything else – no token can have a value for which no brand token exists.

Reviewer: This isn't currently true

> Brand tokens must only be used in common tokens, not in component tokens.

Reviewer: And neither is this 😁

Author: Added a remark.

### Overriding tokens

The value of tokens must not be overridden in an application for the City of Amsterdam.
If you need a different value, create a new token with a different name – and let us know.
This way, the design system remains consistent and predictable.

A custom theme could restyle our components for a different brand.
This is an essential feature of [NL Design System](https://nldesignsystem.nl/) of which we are part.

## Support

Contact us if you have a question, find an issue, or want to contribute.

> Find ways to reach us on [designsystem.amsterdam](https://designsystem.amsterdam).

Reviewer: ```suggestion
Find ways to reach us on [designsystem.amsterdam](https://designsystem.amsterdam/?path=/docs/docs-introduction--docs#send-a-message).
```

Reviewer: True, but currently it sort of feels like: 'Here's the root url, figure it out yourself'. It's not very user centric, imo. 

I thought we had a ticket to add [SUPPORT.md](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/adding-support-resources-to-your-project) to our repo, but I can't find it. Eventually, we should link to that I think, but for now I would prefer the direct link.

<!-- @license CC0-1.0 -->

# Amsterdam Design System: Tokens

This package provides all design tokens from the [Amsterdam Design System](https://designsystem.amsterdam).
Use it to apply its visual design to your website or application.

> The tokens are pure CSS, making them compatible with most if not all technology stacks.

Reviewer: Don't we want to mention some of the other exports we have here? That would make the second part of the sentence a bit more correct. Something like:

```suggestion
The tokens are exported in multiple formats, including CSS custom properties and JSON, making them compatible with most if not all technology stacks.
```

## About

Every design token is a variable representing a single visual design choice.
Tokens exist for colours, various aspects of text, spacing lengths, border widths, the configuration of animations, and more.
The value of every token is valid in CSS.

Some examples:

```css
:root {
  --ams-color-blue: #004699;
  --ams-text-font-weight-bold: 800;
  --ams-border-width-medium: 0.125rem;
  --ams-button-padding-inline: 1.5rem;
  --ams-space-xl: clamp(2.25rem, 2.0625rem + 0.9375vw, 3rem);
  --ams-text-input-invalid-border-color: var(--ams-color-red);
}
```

> Tokens start their name with a prefix of `ams-` in CSS.

Reviewer: ```suggestion
Tokens start their name with a prefix of `--ams-` in CSS.
```

The name of a component token (see below) ends with the property that uses it, e.g. `-font-size`.

All tokens together form a theme that encodes the entire branding of the City of Amsterdam.

## Installation

> Use only this package if you want or need to reference the tokens directly in your project.

Reviewer: ```suggestion
Only use this package if you want or need to reference the tokens directly in your project.
```

Note that our [CSS components](https://www.npmjs.com/package/@amsterdam/design-system-css) and [React components](https://www.npmjs.com/package/@amsterdam/design-system-react) provide more functionality and apply the tokens automatically.
You should use these packages if your application uses React, or if it allows applying our HTML classes.

```sh
npm i @amsterdam/design-system-tokens
```

```sh
pnpm add @amsterdam/design-system-tokens
```

```sh
yarn add @amsterdam/design-system-tokens
```

## Usage in CSS

Tokens are typically used as custom properties in CSS.

### Main stylesheet

This package offers a main stylesheet containing all tokens.
They are declared as global CSS variables through the `:root` selector.

```ts
import "@amsterdam/design-system-tokens/dist/index.css";
```

If that doesn’t work for your project, use `index.theme.css` instead, which uses an `.ams-theme` selector.
Add that class to a root element of your application.

```ts
import "@amsterdam/design-system-tokens/dist/index.theme.css";
```

```html

> <html class="ams-theme">
>   …
> </html>

Reviewer: Setting a class on `html` is a bit unconventional. It is allowed in HTML5 (it wasn't in HTML4), but at best you're styling the `head`, which doesn't have any visual styling 😁 

Setting a global class on `body` makes more sense, imo:

```suggestion
<body class="ams-theme">
  …
</body>
```

```

### Compact mode

Our branding is rather spacious for websites.
The main stylesheet implements this ‘spacious mode’ as the default.
A compact stylesheet is available for applications that require less white space and smaller text.

Note that the compact stylesheet is not independent – it only contains overrides.
Import it after the main stylesheet for the correct result.

```ts
import "@amsterdam/design-system-tokens/dist/index.css";
import "@amsterdam/design-system-tokens/dist/compact.css";
```

> ### Three layers

Reviewer: This section isn't CSS specific anymore, right? Shouldn't we move this after 'Usage in Figma'?

Reviewer: Yeah, I get the trade-off. Ok, let's keep it this way.

The tokens are organised in three layers: brand, common and component tokens.

#### Brand tokens

These express the corporate identity of the City of Amsterdam.
They are our fundamental selection from all possible colours, text characteristics, spacing lengths, border widths, etc.
Examples:

```css
:root {
  --ams-color-primary-red: #ec0000;
  --ams-space-md: 1rem;
  --ams-proportion-wide: 4/3;
  --ams-border-width-lg: 0.1875rem;
}
```

Brand tokens must only be referenced by common tokens; not by component tokens or component stylesheets.

> (We are currently working to apply this rule to our own libraries.)

Reviewer: This is still confusing imo. We seem to be telling consumers of our tokens package that they shouldn't use brand tokens and later on that they shouldn't use component tokens. The common tokens layer doesn't have everything in brand and components yet (and I'm honestly not sure they ever will), so then consumers are out of luck? 

I think we want to encourage consumers to use brand tokens, right? The text you wrote here might become an internal guideline, but I don't think it's relevant for external users.

Author: This may need more thought. I’ve deleted it to allow this PR to go through.

Find the [list of brand tokens](https://github.com/Amsterdam/design-system/tree/main/proprietary/tokens/src/brand/ams) on GitHub.

#### Common tokens

Related components share visual design characteristics.
E.g. all kinds of links have the same colour, and the borders of various form inputs are equally thick.
Common tokens express these relations and streamline future changes.

Design system components use common tokens where possible.
The same goes for custom components that you may create in your application.

```html
<input class="my-input" type="text" />
```

```css
.my-input {
  font-size: var(--ams-common-input-font-size);
  border-width: var(--ams-common-input-border-width);
}
```

> (This is an example; these specific tokens do not exist yet.)

Reviewer: This is also still confusing imo, why not just use common tokens that exist? A link for example?

Reviewer: We can get rid of this line then, no?

```suggestion
```

Author: Committed separately, needed to delete an empty line as well to satisfy the linter.

Find the [list of common tokens](https://github.com/Amsterdam/design-system/tree/main/proprietary/tokens/src/common/ams) on GitHub.

#### Component tokens

Every design system component defines a token for every property that expresses branding and uses it in its stylesheet.

Use these tokens when recreating an existing component to receive the correct values for them – now and in the future.
Do not apply these tokens to other components – components must be independent.

```html
<button class="my-button" type="button">Button label</button>
```

```css
.my-button {
  font-family: var(--ams-button-font-family);
  background-color: var(--ams-button-primary-background-color);
}
```

Find the [list of component tokens](https://github.com/Amsterdam/design-system/tree/main/proprietary/tokens/src/component/ams) on GitHub.

### Overriding tokens

This package allows creating a theme to reuse our components for a different brand.
This is a key feature of [NL Design System](https://nldesignsystem.nl/) of which we are part.

However, websites and applications for the City of Amsterdam must follow the design system as closely as possible.

> We repeat: websites and applications for the City of Amsterdam must follow the design system as closely as possible.

Reviewer: This is kind of funny 😁 For that reason alone, maybe we should keep it. Generally, you would use `em` or `strong` for this though.
